### PR TITLE
Issues/4035 update reader topic models

### PR DESCRIPTION
**Pull request by aerych**
Refs #4035 
Replaces the ReaderTopics model with new more focused models.  This lays the ground work for supporting different kinds of "topics" in the reader, and allows for storing more information about individual topic types. 

To Test:
- Install a copy of 5.6 first, then upgrade to this branch.  Make sure the upgrade works properly. 
- Make sure you can various kinds of topics in the reader.
- Make sure you can follow / unfollow topics
- Make sure the reader shows the Freshly Pressed and the default tag list when not logged into WordPress.com

@astralbodies Would you be up for reviewing this one? 

Needs Review: @astralbodies

### CODE REVIEW

**Comment by astralbodies** [comment]
Looks like this file isn't located in the right spot :smile:

File: WordPress/Classes/Utility/Migrations/20-21/SafeReaderTopicToReaderTopic.m [hunk 1]
```diff
@@ -8,36 +6,11 @@ @implementation SafeReaderTopicToReaderTopic
 
 - (BOOL)beginEntityMapping:(NSEntityMapping *)mapping manager:(NSMigrationManager *)manager error:(NSError **)error
```
Comment (astralbodies) [inline WordPress/Classes/Utility/Migrations/20-21/SafeReaderTopicToReaderTopic.m]: Whoops I hadn't realized we were using Services in migrations. That makes migrations no longer backward compatible.

Comment (aerych) [inline WordPress/Classes/Utility/Migrations/20-21/SafeReaderTopicToReaderTopic.m]: Yeah.   It was a one off and a big no no (for several reasons!). The situation was we needed a way to preserve the current topic when migrating. This was when the topic was being saved as an NSManagedObjectID (horrible, I know).  We're way past that point now and if someone *did* have an app so old it needed to be migrated we'd just let them have the default topic when syncing.  Yay for cleaning up the mess I guess.

**Comment by aerych** [comment]
> Looks like this file isn't located in the right spot

It was swiftly trying to escape?

**Comment by astralbodies** [comment]
Seems to operate properly - will do a more in-depth code review now :smile:

File: WordPress/Classes/Services/ReaderPostService.m [hunk 2]
```diff
@@ -598,13 +598,13 @@ from the result set (deleted, unliked, etc.) rendering the result set empty.
  @param date The date to delete posts earlier than.
  @param topic The ReaderTopic to delete posts from.
```
Comment (astralbodies) [inline WordPress/Classes/Services/ReaderPostService.m]: s/ReaderTopic/ReaderAbstractTopic

File: WordPress/Classes/Services/ReaderPostService.m [hunk 3]
```diff
@@ -739,7 +739,7 @@ - (void)deletePostsFromBlockedSites
  @param topic The `ReaderTopic` to assign to the created posts.
```
Comment (astralbodies) [inline WordPress/Classes/Services/ReaderPostService.m]: s/ReaderTopic/ReaderAbstractTopic

File: WordPress/Classes/Services/ReaderPostService.m [hunk 4]
```diff
@@ -760,7 +760,7 @@ - (NSMutableArray *)makeNewPostsFromRemotePosts:(NSArray *)posts forTopic:(Reade
  @param topic The `ReaderTopic` to assign to the created post.
```
Comment (astralbodies) [inline WordPress/Classes/Services/ReaderPostService.m]: s/ReaderTopic/ReaderAbstractTopic

File: WordPress/Classes/Services/ReaderPostService.h [hunk 5]
```diff
@@ -93,7 +93,7 @@ extern NSString * const ReaderPostServiceErrorDomain;
  */
```
Comment (astralbodies) [inline WordPress/Classes/Services/ReaderPostService.h]: s/ReaderTopic/ReaderAbstractTopic

File: WordPress/Classes/Services/ReaderTopicService.h [hunk 6]
```diff
@@ -73,27 +73,22 @@ extern NSString * const ReaderTopicFreshlyPressedPathCommponent;
 
  @return A `ReaderTopic` instance or nil.
```
Comment (astralbodies) [inline WordPress/Classes/Services/ReaderTopicService.h]: s/ReaderTopic/ReaderAbstractTopic

File: WordPress/Classes/Services/ReaderTopicService.m [hunk 7]
```diff
@@ -449,10 +352,10 @@ - (void)selectTopicWithID:(NSNumber *)topicID
  @param topicName The title of the topic to find in core data.
  @return A matching `ReaderTopic` instance or nil.
```
Comment (astralbodies) [inline WordPress/Classes/Services/ReaderTopicService.m]: s/ReaderTopic/ReaderAbstractTopic

File: WordPress/Classes/Services/ReaderTopicService.m [hunk 8]
```diff
@@ -495,7 +398,7 @@ - (ReaderTopic *)findTopicWithID:(NSNumber *)topicID
  @param dict A `RemoteReaderTopic` object.
  @return A new or updated, but unsaved, `ReaderTopic`.
```
Comment (astralbodies) [inline WordPress/Classes/Services/ReaderTopicService.m]: s/ReaderTopic/ReaderAbstractTopic

File: WordPress/Classes/Services/ReaderTopicService.m [hunk 9]
```diff
@@ -634,12 +585,12 @@ - (NSArray *)allTopics
 }
 
 /**
- Find a specific ReaderTopic by its `path` property.
+ Find a specific ReaderAbstractTopic by its `path` property.
 
  @param path The unique, cannonical path of the topic.
  @return A matching `ReaderTopic` or nil if there is no match.
```
Comment (astralbodies) [inline WordPress/Classes/Services/ReaderTopicService.m]: s/ReaderTopic/ReaderAbstractTopic

File: WordPress/Classes/Models/ReaderDefaultTopic.swift [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import Foundation
+
+@objc public class ReaderDefaultTopic : ReaderAbstractTopic
+{
+    public class var TopicType: String {
```
Comment (astralbodies) [inline WordPress/Classes/Models/ReaderDefaultTopic.swift]: I'm really surprised you're not required to mark these as override!

**Comment by astralbodies** [comment]
Xcode 7 will make you stabby with the Swift data models. Putting the module prefix in the class name will break when we migrate. :cry:

**Comment by astralbodies** [comment]
Everything looks good from my perspective! :shipit:

File: WordPress/Classes/Models/ReaderDefaultTopic.swift (hunk 10 above)
Comment (aerych) [inline WordPress/Classes/Models/ReaderDefaultTopic.swift]: Oops. TopicType isn't declared in the base class otherwise those would have to be overridden.

**Comment by aerych** [comment]
Thanks @astralbodies !!

**Pull request merged**